clippy_lints/src/not_exhaustive_enough.rs

```diff
@@ -0,0 +1,172 @@
+use rustc_hir::{Arm, Expr, ExprKind, FieldPat, MatchSource, Pat, PatKind, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:**
+    /// As suggested in the [non_exhaustive RFC](https://github.com/rust-lang/rfcs/blob/master/text/2008-non-exhaustive.md#unresolved-questions), when using non-exhaustive enums and structs in patterns,
+    /// the lint would warn the user for missing variants or fields despite having a wildcard arm or a rest pattern.
+    ///
+    /// **Why is this bad?**
+    /// When new fields/variants are added by the upstream crate they might go unnoticed.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust
+    /// #[non_exhaustive]
+    /// # enum E {First,Second,Third}
+    /// # let e = E::First;
+    /// // Bad
+    /// match e {
+    ///         E::First => {}
+    ///         E::Second => {}
+    ///         _ => {}
+    ///     }
+    /// // Good
+    /// match e {
+    ///         E::First => {}
+    ///         E::Second => {}
+    ///         E::Third => {}
+    ///     }
+    /// ```
+    pub NOT_EXHAUSTIVE_ENOUGH,
+    pedantic,
+    "warn the user for missing variants or fields despite having a wildcard arm or a rest pattern"
+}
+
+declare_lint_pass!(NotExhaustiveEnough => [NOT_EXHAUSTIVE_ENOUGH]);
+
+impl<'tcx> LateLintPass<'tcx> for NotExhaustiveEnough {
+    fn check_pat(&mut self, cx: &LateContext<'tcx>, pat: &'tcx Pat<'_>) {
+        if_chain! {
+            if let PatKind::Struct(_, ref field_pats, ref rest_pat) = &pat.kind;
+            if let ty::Adt(adt_def, _) = cx.typeck_results().pat_ty(pat).kind();
+            if adt_def.is_struct();
+            if is_struct_not_exhaustive(adt_def);
+            if *rest_pat;
+            if !field_pats.is_empty();
+            if let Some(variant) = get_variant(adt_def);
+            if let field_defs = &variant.fields;
+            then
+            {
+                check_struct_pat(cx, pat, field_pats, field_defs);
+            }
+        }
+    }
+
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if_chain! {
+            if let ExprKind::Match(e, ref arms, MatchSource::Normal) = expr.kind;
+            if !arms.is_empty();
+            if let ExprKind::Path(..) = e.kind;
+            if let ty::Adt(adt_def, _) = cx.typeck_results().expr_ty(e).kind();
+            if adt_def.is_enum();
+            if is_enum_not_exhaustive(adt_def);
+            then
+            {
+                check_path_pat(cx, arms, e);
+            }
+        }
+    }
+}
+
+fn check_path_pat<'tcx>(cx: &LateContext<'tcx>, arms: &[Arm<'_>], e: &'tcx Expr<'_>) {
+    let missing_variants = get_missing_variants(cx, arms, e);
+    span_lint_and_sugg(
+        cx,
+        NOT_EXHAUSTIVE_ENOUGH,
+        e.span,
+        "Enum not exhaustive enough",
```

Comment:
```suggestion
        "enum match is not exhaustive enough",
```
We generally avoid capitalizing the first letter